core/src/main/java/com/backblaze/b2/json/B2JsonByteArrayOutputStream.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright 2021, Backblaze Inc. All Rights Reserved.
+ * License https://www.backblaze.com/using_b2_code.html
+ */
+package com.backblaze.b2.json;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.UnsupportedEncodingException;
+import java.util.Arrays;
+
+/**
+ * A B2Json version of ByteArrayOutputStream that throws IOException (as opposed to OutOfMemoryError)
+ * when new capacity would be over the maximum limit (Integer.MAX_VALUE - 8). Everything else works
+ * the same way as the ByteArrayOutputStream from the standard Java version.
+ *
+ * THREAD-SAFE
+ */
+public class B2JsonByteArrayOutputStream extends OutputStream {
+
+    /**
+     * The maximum size of array to allocate. Reserve some header words
+     * in an array. Attempts to allocate larger arrays will result in
+     * IOException: Requested array size exceeds maximum limit.
+     */
+    private static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
```

Comment:
How did you arrive at `Integer.MAX_VALUE - 8` being the proper maximum that we should support?

Should this be configurable? Integer.MAX_VALUE is pretty large, it may be too large for many applications that don't configure huge heaps for Java

Reply:
this value is used in the original version of ByteArrayOutputStream. There is not much detail about it in that version. I'd agree this definitely looks to be too large since we typically get into OOM error without triggering IOException (I did manage to get IOException with 2 GB heap). But I don't know what would be a reasonable value for this yet.